### PR TITLE
br: merge feature branch of ebs-based-br 

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #35306

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/restore/data.go [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+)
+
+// recover the tikv cluster
+// 1. read all meta data from tikvs
+// 2. Assemble region meta and make recovery plan
+// 3. send the recover plan and the wait tikv to apply, in waitapply, all assigned region leader will check and apply log to the last log
+// 4. send the resolvedTs to tikv for deleting data.
+func RecoverData(ctx context.Context, resolvedTs uint64, allStores []*metapb.Store, tls *tls.Config, dumpRegionInfo bool) (int, error) {
+
+	numOfTiKVs := len(allStores)
```
Comment (3pointer) [inline br/pkg/restore/data.go:32]: it seems we don't need this variable `numOfTiKVs`. change it by `len(allStores)` is enough?

File: br/pkg/restore/data.go [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+	err := recovery.ReadRegionMeta(ctx, numOfTiKVs, allStores)
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+
+	if dumpRegionInfo {
+		recovery.dumpRegionInfo()
+	}
+
+	totalRegions := recovery.getTotalRegions()
```
Comment (3pointer) [inline br/pkg/restore/data.go:44]: please add some logs about this information.

File: br/pkg/restore/data.go [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+	for storeId, plan := range recovery.recoveryPlan {
+		storeAddr := getStoreAddress(allStores, storeId)
+		tikvClient, err := recovery.newTiKVRecoveryClient(ectx, storeAddr)
+		if err != nil {
+			log.Error("create tikv client failed", zap.Uint64("storeID", storeId))
+			return errors.Trace(err)
+		}
+		cmd := plan
+		workers.ApplyWithIDInErrorGroup(eg, func(id uint64) error {
+			stream, err := tikvClient.RecoverCmd(ectx)
```
Comment (3pointer) [inline br/pkg/restore/data.go:265]: RecoverCmd is not a clear name, do we have chance to rename it to `RecoverRegions` of `RecoverRegionMetas`?

File: br/pkg/restore/data.go [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+
+	if dumpRegionInfo {
+		recovery.dumpRegionInfo()
+	}
+
+	totalRegions := recovery.getTotalRegions()
+
+	recovery.makeRecoveryPlan()
+
+	err = recovery.RecoverRegions(ctx, allStores)
```
Comment (3pointer) [inline br/pkg/restore/data.go:48]: please add some logs when RecoverRegions finished.

File: br/pkg/restore/data.go [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+	if err != nil {
+		return totalRegions, errors.Trace(err)
+	}
+
+	err = recovery.ResolveData(ctx, allStores, resolvedTs)
```
Comment (3pointer) [inline br/pkg/restore/data.go:53]: ditto

File: br/pkg/restore/data.go [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+	for i := 0; i < totalTiKVs; i++ {
+		i := i
+		storeId := allStores[i].GetId()
+		tikvClient, err := recovery.newTiKVRecoveryClient(ectx, allStores[i].GetAddress())
+		if err != nil {
+			log.Error("create tikv client failied", zap.Uint64("storeID", storeId))
+			return errors.Trace(err)
+		}
+		log.Debug("read meta from tikv", zap.String("tikv address", allStores[i].GetAddress()))
+		workers.ApplyWithIDInErrorGroup(eg, func(id uint64) error {
```
Comment (3pointer) [inline br/pkg/restore/data.go:160]: seems `ApplyOnErrorGroup` is enough?

File: br/pkg/restore/data.go [hunk 7]
```diff
@@ -0,0 +1,503 @@
...
+
+	for storeId, plan := range recovery.recoveryPlan {
+		storeAddr := getStoreAddress(allStores, storeId)
+		tikvClient, err := recovery.newTiKVRecoveryClient(ectx, storeAddr)
+		if err != nil {
+			log.Error("create tikv client failed", zap.Uint64("storeID", storeId))
+			return errors.Trace(err)
+		}
+		cmd := plan
+		workers.ApplyWithIDInErrorGroup(eg, func(id uint64) error {
```
Comment (3pointer) [inline br/pkg/restore/data.go:264]: seems ApplyOnErrorGroup is enough?

File: br/pkg/restore/data.go [hunk 8]
```diff
@@ -0,0 +1,503 @@
...
+	log.Debug("resolved kv data started", zap.Uint64("resolve-ts", resolvedTs))
+	for _, store := range allStores {
+		storeAddr := getStoreAddress(allStores, store.Id)
+		tikvClient, err := recovery.newTiKVRecoveryClient(ectx, storeAddr)
+		if err != nil {
+			log.Error("create tikv client failed", zap.String("ip", storeAddr))
+			return errors.Trace(err)
+		}
+
+		workers.ApplyWithIDInErrorGroup(eg, func(id uint64) error {
```
Comment (3pointer) [inline br/pkg/restore/data.go:310]: ditto

File: br/pkg/task/restore.go [hunk 9]
```diff
@@ -71,6 +71,8 @@ const (
 	TableRestoreCmd = "Table Restore"
 	PointRestoreCmd = "Point Restore"
 	RawRestoreCmd   = "Raw Restore"
+
+	EBSMetaRestoreCmd = "EBS meta restore"
```
Comment (3pointer) [inline br/pkg/task/restore.go:75]:
```suggestion
	EBSMetaRestoreCmd = "EBS Meta Restore"
```

File: br/pkg/task/restore_data.go [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+	RestoreDataCmd = "Restore Data"
+)
+
+const (
+	flagDumpRegionInfo = "dump-region-info"
+)
+
+// DefineRestoreDataFlags defines common flags for the restore command.
+func DefineRestoreDataFlags(command *cobra.Command) {
+	command.Flags().Bool(flagDryRun, false, "don't access to aws environment if set to true")
```
Comment (3pointer) [inline br/pkg/task/restore_data.go:37]: do we need dry-run any more? 🤔

File: br/pkg/task/common.go [hunk 11]
```diff
@@ -75,11 +76,16 @@ const (
 	// flagEnableOpenTracing is whether to enable opentracing
 	flagEnableOpenTracing = "enable-opentracing"
 	flagSkipCheckPath     = "skip-check-path"
-	flagWithSysTable      = "with-sys-table"
+	flagDryRun            = "dry-run"
```
Comment (3pointer) [inline br/pkg/task/common.go:79]: this can be removed

File: br/pkg/restore/util_test.go [hunk 12]
```diff
@@ -291,7 +293,7 @@ func TestPaginateScanRegion(t *testing.T) {
 	_, err = split.PaginateScanRegion(ctx, NewTestClient(stores, regionMap, 0), regions[1].Region.EndKey, regions[5].Region.EndKey, 3)
 	require.Error(t, err)
 	require.True(t, berrors.ErrPDBatchScanRegion.Equal(err))
-	require.Regexp(t, ".*region endKey not equal to next region startKey.*", err.Error())
+	require.Regexp(t, ".*region endKey not equal to next region StartKey.*", err.Error())
```
Comment (3pointer) [inline br/pkg/restore/util_test.go:296]:
```suggestion
	require.Regexp(t, ".*region endKey not equal to next region startKey.*", err.Error())
```

File: br/pkg/task/common.go (hunk 11 above)
Comment (D3Hunter) [reply br/pkg/task/common.go:79]: pitr is using it

File: br/pkg/task/stream.go [hunk 13]
```diff
@@ -59,7 +59,6 @@ import (
 
 const (
 	flagYes              = "yes"
-	flagDryRun           = "dry-run"
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply br/pkg/task/stream.go:62]: it's moved into `common.go`

**Review by fengou1 — APPROVED** [review]

File: br/pkg/task/backup.go [hunk 14]
```diff
@@ -77,6 +78,13 @@ type BackupConfig struct {
 	IgnoreStats      bool          `json:"ignore-stats" toml:"ignore-stats"`
 	UseBackupMetaV2  bool          `json:"use-backupmeta-v2"`
 	CompressionConfig
+
+	// for ebs-based backup
+	FullBackupType      FullBackupType `json:"full-br-type" toml:"full-br-type"`
```
Comment (3pointer) [inline br/pkg/task/backup.go:83]:
```suggestion
	FullBackupType      FullBackupType `json:"full-backup-type" toml:"full-backup-type"`
```
should we change this?

**Review by 3pointer — APPROVED** [review]

**Comment by D3Hunter** [comment]
ci success https://ci.pingcap.net/blue/organizations/jenkins/br_ghpr_unit_and_integration_test/detail/br_ghpr_unit_and_integration_test/13334/pipeline

**Pull request merged**